File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/workspace/ProjectDetectionTest.kt
package com.tyhp.lang.workspace

import com.tyhp.lang.lsp.ProjectFileFs
import com.tyhp.lang.lsp.ResolveProjectFileOptions
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class ProjectDetectionTest {
    private fun fsFrom(files: Map<String, String>): ProjectFileFs {
        return object : ProjectFileFs {
            override fun exists(path: String): Boolean = files.containsKey(path)
            override fun isDirectory(path: String): Boolean = files[path] == "dir"
        }
    }

    private fun posix(configured: String, roots: List<String>, files: Map<String, String>): WorkspaceSnapshot {
        return detectWorkspaceProject(
            ResolveProjectFileOptions(
                configuredPath = configured,
                contentRoots = roots,
                join = { dir, name -> "$dir/$name".replace("//", "/") },
                resolve = { root, rel ->
                    if (rel.startsWith("/")) rel else "$root/$rel".replace("//", "/")
                },
                isAbsolute = { it.startsWith("/") },
                fs = fsFrom(files),
            ),
        )
    }

    @Test
    fun `empty tyhp projectPath does not treat content-root tyhp json as the owner`() {
        val snapshot = posix("", listOf("/empty", "/app"), mapOf("/app/tyhp.json" to "file"))
        assertNull(snapshot.projectFilePath)
        assertEquals("not in a Tyhp project", projectStatusLabel(snapshot))
    }

    @Test
    fun `honors tyhp projectPath override as a file path`() {
        val snapshot = posix(
            "/proj/custom.json",
            listOf("/ws"),
            mapOf(
                "/proj/custom.json" to "file",
                "/ws/tyhp.json" to "file",
            ),
        )
        assertEquals("/proj/custom.json", snapshot.projectFilePath)
        assertEquals("proj", snapshot.projectName)
    }

    @Test
    fun `returns no project when tyhp json is absent`() {
        val snapshot = posix("  ", listOf("/ws"), emptyMap())
        assertNull(snapshot.projectFilePath)
        assertEquals("not in a Tyhp project", projectStatusLabel(snapshot))
    }

    @Test
    fun `project display name is the directory containing tyhp json`() {
        val snapshot = snapshotFromProjectFile("/Users/me/code/demo/tyhp.json")
        assertEquals("demo", snapshot.projectName)
        assertEquals("demo", projectStatusLabel(snapshot))
    }

    @Test
    fun `content root prefers the longest root that contains the file`() {
        assertEquals(
            "/ws/app",
            contentRootForPath("/ws/app/src/Main.tyhp", listOf("/ws", "/ws/app")),
        )
        assertEquals("/ws", contentRootForPath(null, listOf("/ws", "/other")))
        assertNull(contentRootForPath(null, emptyList()))
    }
}

File: tyhp-lang/vscode/src/workspace/projectDetection.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import {
    detectWorkspaceProject,
    projectStatusLabel,
    snapshotFromProjectFile,
} from "./projectDetection";

function fsFrom(files: Record<string, "file" | "dir">) {
    return {
        existsSync(target: string): boolean {
            return Object.prototype.hasOwnProperty.call(files, target);
        },
        isDirectory(target: string): boolean {
            return files[target] === "dir";
        },
    };
}

const posix = {
    join: (a: string, b: string) => `${a.replace(/\/$/, "")}/${b}`,
    resolve: (...parts: string[]) => {
        let out = "";
        for (const part of parts) {
            if (part.startsWith("/")) {
                out = part;
            } else {
                out = `${out.replace(/\/$/, "")}/${part}`;
            }
        }
        return out;
    },
    isAbsolute: (target: string) => target.startsWith("/"),
    dirname: (target: string) => {
        const i = target.lastIndexOf("/");
        return i <= 0 ? "/" : target.slice(0, i);
    },
    basename: (target: string) => {
        const i = target.lastIndexOf("/");
        return i < 0 ? target : target.slice(i + 1);
    },
};

test("empty tyhp.projectPath does not treat workspace-root tyhp.json as the owner", () => {
    const snapshot = detectWorkspaceProject(
        {
            configuredPath: "",
            workspaceRoots: ["/empty", "/app"],
            join: posix.join,
            resolve: posix.resolve,
            isAbsolute: posix.isAbsolute,
            fs: fsFrom({
                "/app/tyhp.json": "file",
            }),
        },
        posix
    );
    assert.equal(snapshot.projectFilePath, undefined);
    assert.equal(projectStatusLabel(snapshot), "not in a Tyhp project");
});

test("honors tyhp.projectPath override as a file path", () => {
    const snapshot = detectWorkspaceProject(
        {
            configuredPath: "/proj/custom.json",
            workspaceRoots: ["/ws"],
            join: posix.join,
            resolve: posix.resolve,
            isAbsolute: posix.isAbsolute,
            fs: fsFrom({
                "/proj/custom.json": "file",
                "/ws/tyhp.json": "file",
            }),
        },
        posix
    );
    assert.equal(snapshot.projectFilePath, "/proj/custom.json");
    assert.equal(snapshot.projectName, "proj");
});

test("returns not in a Tyhp project when tyhp.json is absent", () => {
    const snapshot = detectWorkspaceProject(
        {
            configuredPath: "",
            workspaceRoots: ["/ws"],
            join: posix.join,
            resolve: posix.resolve,
            isAbsolute: posix.isAbsolute,
            fs: fsFrom({}),
        },
        posix
    );
    assert.equal(snapshot.projectFilePath, undefined);
    assert.equal(projectStatusLabel(snapshot), "not in a Tyhp project");
});

test("project display name is the directory containing tyhp.json", () => {
    const snapshot = snapshotFromProjectFile("/Users/me/code/demo/tyhp.json", posix);
    assert.equal(snapshot.projectName, "demo");
    assert.equal(projectStatusLabel(snapshot), "demo");
});
